### DISCUSSION TITLE
interactive shell question

### DISCUSSION THREAD

**Discussion by 2ctony** [ParallelSSH/ssh2-python#175] [Q&A] [answered]
I have looked at the code examples and still struggling w/ a proper interactive shell example I can use for running multiple commands and capturing stdout for each.   I understand .read will block indefinitely since there is no EOF.
So how does one loop using interactive shell with out blocking?  
I tried sleeps to wait for stdout to show up, set_blocking etc
I see polling mentioned in another thread, but I didn't see an example.
In paramiko I would looping using recv_ready().  
Anyways, I would greatly appreciate a point in the right direction.

Tks,
Tony
 
<init>
channel = session.open_session()
channel.pty()
shell = channel.shell()

channel.write('some command \n') 
read this stdout of variable length 
channel.write('some command \n') 
read 
etc
channel.close()

Btw have also tried non interactive w/ execute and facing a different problem.  Opening a new channel .open_session() for the next command results in a SocketRecvError.  Assuming the socket is close with out digging?  
I can go either direction so long as I can get something working.

**Comment by pkittenis** [comment, marked as answer]
Hi there,

See section on interactive shells in documentation of [parallel-ssh](https://github.com/ParallelSSH/parallel-ssh).

https://parallel-ssh.readthedocs.io/en/latest/advanced.html#running-commands-on-shells

Eg for single clients:
```
client = SSHClient(<..>)

cmd = 'echo me'
with client.open_shell() as shell:
    shell.run(cmd)
print(list(shell.stdout))
print(shell.exit_code)
```

**Reply by 2ctony** [reply]
Thank you.
